ast/compile.go

```diff
@@ -1728,6 +1735,218 @@ func (c *Compiler) rewriteDynamicTerms() {
+
+				var metadataChainVar Var
+				if regoMetadataChainCalled {
+					// Create and inject metadata chain for rule
+
+					chain, err := createMetadataChain(c.annotationSet.Chain(rule))
+					if err != nil {
+						c.err(err)
+						return false
+					}
+
+					f := newEqualityFactory(c.localvargen)
+					eq := f.Generate(chain)
+					metadataChainVar = eq.Operands()[0].Value.(Var)
+					body.Append(eq)
+				}
+
+				var metadataRuleVar Var
+				if regoMetadataRuleCalled {
+					// Create and inject metadata for rule
+
+					var metadataRuleTerm *Term
+
+					a := getPrimaryRuleAnnotations(c.annotationSet, rule)
+					if a != nil {
+						annotObj, err := a.toObject()
+						if err != nil {
+							c.err(err)
+							return false
+						}
+						metadataRuleTerm = NewTerm(*annotObj)
+					} else {
+						// If rule has no annotations, assign an empty object
+						metadataRuleTerm = ObjectTerm()
+					}
+
+					f := newEqualityFactory(c.localvargen)
+					eq := f.Generate(metadataRuleTerm)
+					metadataRuleVar = eq.Operands()[0].Value.(Var)
+					body.Append(eq)
+				}
+
+				for _, expr := range rule.Body {
+					body.Append(expr)
+				}
+				rule.Body = body
+
+				vis := func(b Body) bool {
+					for _, err := range rewriteRegoMetadataCalls(metadataChainVar, metadataRuleVar, b, &c.RewrittenVars) {
+						c.err(err)
+					}
+					return false
+				}
+				WalkBodies(rule.Head, vis)
+				WalkBodies(rule.Body, vis)
+			}
+
+			return false
+		})
+	}
+}
+
+func getPrimaryRuleAnnotations(as *AnnotationSet, rule *Rule) *Annotations {
+	annots := as.GetRuleScope(rule)
+
+	if len(annots) == 0 {
+		return nil
+	}
+
+	if len(annots) > 1 {
+		// Sort by annotation location; chain must start with annotations declared closest to rule, then going outward
+		sort.SliceStable(annots, func(i, j int) bool {
+			return annots[i].Location.Compare(annots[j].Location) > 0
+		})
+	}
+
+	return annots[0]
+}
+
+func rewriteRegoMetadataCalls(metadataChainVar Var, metadataRuleVar Var, body Body, rewrittenVars *map[Var]Var) Errors {
+	var errs Errors
+
+	WalkClosures(body, func(x interface{}) bool {
+		switch x := x.(type) {
+		case *ArrayComprehension:
+			errs = rewriteRegoMetadataCalls(metadataChainVar, metadataRuleVar, x.Body, rewrittenVars)
+		case *SetComprehension:
+			errs = rewriteRegoMetadataCalls(metadataChainVar, metadataRuleVar, x.Body, rewrittenVars)
+		case *ObjectComprehension:
+			errs = rewriteRegoMetadataCalls(metadataChainVar, metadataRuleVar, x.Body, rewrittenVars)
+		case *Every:
+			errs = rewriteRegoMetadataCalls(metadataChainVar, metadataRuleVar, x.Body, rewrittenVars)
+		}
+		return true
+	})
+
+	for i := range body {
+		expr := body[i]
+		var metadataVar Var
+		// We're replacing a function call, that we then need to replace in error reporting
+		var originalVar Var
+
+		if isRegoMetadataChainCall(expr) {
+			metadataVar = metadataChainVar
+			originalVar = Var(fmt.Sprintf("%s()", RegoMetadataChain.Ref().String()))
```

Comment:
Hhmm! So `originalVar` in this case isn't _really_ a var at all, is it? But we make-believe it was by creating a Var with the name of the `rego.metadata.*` function? What are the errors that we do this for? Is there perhaps some other way to get them to be more userfriendly?

Reply:
Let's add a to-do comment maybe? To revisit this later...